TestDistributedFaultTolerantEngineOnlyQueries: tried to release more than allocated

```
2022-04-04T09:37:22.484-0500	ERROR	pool-5-thread-2	io.trino.testng.services.ProgressLoggingListener	[TEST FAILURE] io.trino.tests.TestDistributedFaultTolerantEngineOnlyQueries.testUnionRequiringCoercion; (took: 1.1 seconds)
java.lang.AssertionError: Execution of 'actual' query failed: SELECT * FROM (VALUES (2, 2)) UNION SELECT * FROM (VALUES (1, 1.0))
	at org.testng.Assert.fail(Assert.java:83)
	at io.trino.testing.QueryAssertions.assertQuery(QueryAssertions.java:150)
	at io.trino.testing.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at io.trino.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:248)
	at io.trino.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:243)
	at io.trino.testing.AbstractTestEngineOnlyQueries.testUnionRequiringCoercion(AbstractTestEngineOnlyQueries.java:5721)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: tried to release more than allocated (1073741824 vs null) for node 638624cd-df26-421a-b87c-d818c0b2d9a0
	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:122)
	at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:527)
	at io.trino.testing.QueryAssertions.assertQuery(QueryAssertions.java:147)
	... 17 more
	Suppressed: java.lang.Exception: SQL: SELECT * FROM (VALUES (2, 2)) UNION SELECT * FROM (VALUES (1, 1.0))
		at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:530)
		... 18 more
Caused by: com.google.common.base.VerifyException: tried to release more than allocated (1073741824 vs null) for node 638624cd-df26-421a-b87c-d818c0b2d9a0
	at com.google.common.base.Verify.verify(Verify.java:436)
	at io.trino.execution.scheduler.BinPackingNodeAllocatorService.lambda$updateAllocatedMemory$1(BinPackingNodeAllocatorService.java:238)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1908)
	at io.trino.execution.scheduler.BinPackingNodeAllocatorService.updateAllocatedMemory(BinPackingNodeAllocatorService.java:235)
	at io.trino.execution.scheduler.BinPackingNodeAllocatorService$BinPackingNodeLease.release(BinPackingNodeAllocatorService.java:299)
	at io.trino.execution.scheduler.FaultTolerantStageScheduler.updateTaskStatus(FaultTolerantStageScheduler.java:545)
	at io.trino.execution.scheduler.FaultTolerantStageScheduler.lambda$startTask$7(FaultTolerantStageScheduler.java:372)
	at io.trino.execution.StateMachine.fireStateChangedListener(StateMachine.java:241)
	at io.trino.execution.StateMachine.lambda$fireStateChanged$0(StateMachine.java:233)
	at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:80)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java
+++ b/core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java
@@ -211,12 +211,12 @@ public class BinPackingNodeAllocatorService
                 case RESERVED:
                     InternalNode reservedNode = result.getNode().orElseThrow();
                     fulfilledAcquires.add(pendingAcquire.getLease());
+                    updateAllocatedMemory(reservedNode, pendingAcquire.getMemoryLease());
                     pendingAcquire.getFuture().set(reservedNode);
-                    if (!pendingAcquire.getFuture().isCancelled()) {
-                        updateAllocatedMemory(reservedNode, pendingAcquire.getMemoryLease());
-                    }
-                    else {
-                        // request was cancelled in the meantime
+                    if (pendingAcquire.getFuture().isCancelled()) {
+                        // completing future was unsuccessful - request was cancelled in the meantime
+                        pendingAcquire.getLease().deallocateMemory(reservedNode);
+
                         fulfilledAcquires.remove(pendingAcquire.getLease());
 
                         // run once again when we are done
@@ -316,6 +316,7 @@ public class BinPackingNodeAllocatorService
     {
         private final SettableFuture<InternalNode> node = SettableFuture.create();
         private final AtomicBoolean released = new AtomicBoolean();
+        private final AtomicBoolean memoryDeallocated = new AtomicBoolean();
         private final long memoryLease;
         private final AtomicReference<TaskId> taskId = new AtomicReference<>();
 
@@ -369,13 +370,20 @@ public class BinPackingNodeAllocatorService
             if (released.compareAndSet(false, true)) {
                 node.cancel(true);
                 if (node.isDone() && !node.isCancelled()) {
-                    updateAllocatedMemory(getFutureValue(node), -memoryLease);
+                    deallocateMemory(getFutureValue(node));
                     wakeupProcessPendingAcquires();
                     checkState(fulfilledAcquires.remove(this), "node lease %s not found in fulfilledAcquires %s", this, fulfilledAcquires);
                 }
             }
             else {
                 throw new IllegalStateException("Node " + node + " already released");
+            }
+        }
+
+        public void deallocateMemory(InternalNode node)
+        {
+            if (memoryDeallocated.compareAndSet(false, true)) {
+                updateAllocatedMemory(node, -memoryLease);
             }
         }
     }

--- a/core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java
+++ b/core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java
@@ -580,6 +580,20 @@ public class TestBinPackingNodeAllocator
         }
     }
 
+    @Test
+    public void testStressAcquireRelease()
+    {
+        InMemoryNodeManager nodeManager = testingNodeManager(basicNodesMap(NODE_1));
+        setupNodeAllocatorService(nodeManager, DataSize.of(4, GIGABYTE));
+
+        try (NodeAllocator nodeAllocator = nodeAllocatorService.getNodeAllocator(SESSION)) {
+            for (int i = 0; i < 10_000_000; ++i) {
+                NodeAllocator.NodeLease lease = nodeAllocator.acquire(REQ_32);
+                lease.release();
+            }
+        }
+    }
+
     private TaskId taskId(int partition)
     {
         return new TaskId(new StageId("test_query", 0), partition, 0);

